### PR TITLE
Trigger new deployment chain via piped api

### PR DESCRIPTION
**Pull request by khanhtc1202**
**What this PR does / why we need it**:

This PR provides:
- logic to create a new deployment chain on trigger application which contains `spec.postSync.chain` configuration
- add grpc CreateDeploymentChain which add deployment chain model to control-plane datastore and make command to trigger all applications of the chain

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/api/service/pipedservice/service.proto [hunk 1]
```diff
@@ -477,3 +481,16 @@ message ReportUnregisteredApplicationConfigurationsRequest {
...
 message ReportUnregisteredApplicationConfigurationsResponse {
 }
+
+message CreateDeploymentChainRequest {
+    message ApplicationsFilter {
+        string app_name = 1;
+        string app_kind = 2;
+        map<string,string> app_labels = 3;
+    }
+
```
Comment (nghialv) [inline pkg/app/api/service/pipedservice/service.proto:491]: How about using the same `Node` as it is?

File: pkg/app/api/grpcapi/piped_api.go [hunk 2]
```diff
@@ -963,6 +966,93 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
...
+
+		apps = append(apps, nodeApps...)
+		chainNodes = append(chainNodes, &model.DeploymentChainNode{
+			Index:    int32(i + 1),
+			Runnable: false,
+		})
+	}
+
+	dc := model.DeploymentChain{
+		Id:        uuid.New().String(),
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:1022]: Maybe we should add this chain ID to the deployment model.

File: pkg/app/api/grpcapi/piped_api.go [hunk 3]
```diff
@@ -963,6 +967,93 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
...
+			PipedId:       app.PipedId,
+			ApplicationId: app.Id,
+			ProjectId:     app.ProjectId,
+			Commander:     dc.Id,
+			// TODO: Add new command type in order to separate this in chain sync command with other commands.
+			Type: model.Command_SYNC_APPLICATION,
+			SyncApplication: &model.Command_SyncApplication{
+				ApplicationId: app.Id,
+				SyncStrategy:  model.SyncStrategy_AUTO,
+			},
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:1046]: Maybe we need a way to distinguish from the normal `SYNC` command emitted via the web to have a better summary for its deployment.
https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/trigger/trigger.go#L259

File: pkg/app/api/grpcapi/piped_api.go (hunk 2 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1023]: You're right, we need that as a reference to display which chain this deployment belongs to in the deployment detailed page 👍

File: pkg/app/api/service/pipedservice/service.proto (hunk 1 above)
Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: 👍

File: pkg/app/api/grpcapi/piped_api.go (hunk 3 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1046]: We are on the same page 👍  My plan is to make a new command type (such as `IN_CHAIN_SYNC_APPLICATION`) and filter out all commands with type prefix `IN_CHAIN` and handler it with `OnChainDeterminer` instead of `OnCommandDeterminer`. Also, the new command kind needs to have: deploymentChainId and NodeIndex at least so that the pipedx (who handles the command) will be able to update this deployment chain model its deployment state correctly. That is the thing I'm going to do with that TODO comment.

File: pkg/app/api/service/pipedservice/service.proto (hunk 1 above)
Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: tbh, I'm just want to rename the `Node` to `ApplicationsFilter` since those things are from the `spec.postSync.chain` configuration and at the time we load it from the configuration file, it's kind of filter used to find applications more than `node` or `applications`, the `node` word come when we create deployment chain, and `applications` come from users side, just to notify users that the filters written there used to find out applications for the chain.

File: pkg/app/api/grpcapi/piped_api.go (hunk 2 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1023]: addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/050e72b0f1063021c77c4855c113db657e515cae

Comment (nghialv) [reply pkg/app/api/grpcapi/piped_api.go:1023]: The first deployment needs this ID as well. Do you think that we should create it by the first Piped?

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1023]: I see, in such case, the first piped will create deploymentChain object and register it to the control-plane just as we do in `trigger.triggerDeployment` currently, right?

**Comment by nakabonne** [comment]
Looks like `web-test` has been failed.

File: pkg/app/api/grpcapi/piped_api.go (hunk 2 above)
Comment (nghialv) [reply pkg/app/api/grpcapi/piped_api.go:1023]: How it can create the whole deployment chain object while it doesn't have enough information of applications from other Piped?
Do we have any good ideas? 🤔

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1023]: No, we don't have to create a whole deployment chain object, just the deploymentChainId is enough, we still submit the filters used to find application and make command as we do in this PR, the deploymentChain object created by the piped just to enable us to use the deploymentChainId to trigger the first deployment of the chain, everything else in this PR has remained.

File: pkg/app/api/service/pipedservice/service.proto (hunk 1 above)
Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: Ok. I get your points.

So it is time to find some good names for all those things: name in the config, name in gRPC request, name in chain model. 🤔 

How do you think about this?


``` go

// config

type DeploymentChain struct {
	Applications []ChainApplicationMatcher`json:"applications"`
	// Conditions *ChainTriggerCondition `json:"conditions,omitempty"`
}

type ChainApplicationMatcher struct {
	Name   string            `json:"name"`
	Kind   string            `json:"kind"`
	Labels map[string]string `json:"labels"`
}

// gRPC request

use the same name with config
type CreateDeploymentChainRequest struct {
  ID string
  Applications []ChainApplicationMatcher
}

type ChainApplicationMatcher struct {
  ...
}

// model in Proto format

type DeploymentChain struct {
    ID string
    Blocks []ChainBlock
}

type ChainBlock struct {
    Nodes []ChainNode
}

type ChainNode struct {
     Application *ChainApplicationRef
     Deployment *ChainDeploymentRef
     Runnable bool
}

// Just needed information to render on the web.
type ChainApplicationRef struct {
   AppID string
   AppName string
}

// Just needed information to render on the web.
type ChainDeploymentRef struct {
  DeploymentID string
  ...
  CompletedAt int64
}

```

Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: `ChainBlock` or `ChainGroup` (Maybe we should use ChainBlock to indicate that we are using Blockchain technology in our project. 🤣 )

File: pkg/app/api/grpcapi/piped_api.go (hunk 2 above)
Comment (nghialv) [reply pkg/app/api/grpcapi/piped_api.go:1023]: 👍

File: pkg/app/api/service/pipedservice/service.proto (hunk 1 above)
Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: - 
```golang
// use the same name with config
type CreateDeploymentChainRequest struct {
  ApplicationMatcher {
  }

  ID string
}
```
The CreateDeploymentChainRequest should not contain the applications field, since we will apply the "matcher", find applications, and make sync command for them in control-plane (piped which is not in charge for those applications should not fetch the application info as well so do it in control-plane is better)

- "filter" or "matcher" are both good to me so using "matcher" is fine 🤣 

other things are as same as my image so 🙌

Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: > The CreateDeploymentChainRequest should not contain the applications field, since we will apply the "matcher", find applications

I don't get this. That is the matcher list `  Applications []ApplicationMatcher`. Don't we need them to find apps in the control plane?

Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: Ah sorry, the field named `Applications` make me get it wrong 🙏  Should be `ApplicationsMatchers`
 🤔

Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]:
```golang
type ChainBlock struct {
    Nodes []ChainNode
    Runnable bool
}

type ChainNode struct {
     Application *ChainApplicationRef
     Deployment *ChainDeploymentRef
     CompletedAt int64
}
```
btw, I feel like this, the `Runnable` used by the previous block to unlock this node, and the current block node update its CompletedAt and once all nodes in a block completed, the next block is marked as runnable

Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: Updated! 👍 

``` go

// config

type DeploymentChain struct {
	ApplicationMatchers []ChainApplicationMatcher`json:"applications"`
	// Conditions *ChainTriggerCondition `json:"conditions,omitempty"`
}

type ChainApplicationMatcher struct {
	Name   string            `json:"name"`
	Kind   string            `json:"kind"`
	Labels map[string]string `json:"labels"`
}

// gRPC request

// use the same name with config
type CreateDeploymentChainRequest struct {
  ApplicationMatcher {
  }

  ID string
  ApplicationMatchers []ApplicationMatcher
}

// model in Proto format

type DeploymentChain struct {
    ID string
    Blocks []ChainBlock
}

type ChainBlock struct {
    Nodes []ChainNode
}

type ChainNode struct {
     Application *ChainApplicationRef
     Deployment *ChainDeploymentRef
}

// Just needed information to render on the web.
type ChainApplicationRef struct {
   AppID string
   AppName string
}

// Just needed information to render on the web.
// Basically we just need the "status" and ID to link to the actual one.
type ChainDeploymentRef struct {
  DeploymentID string
  ...
  CompletedAt int64
}

```

Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: About `runnable` I am thinking that maybe we don't need it.
Because in case of having that field, who will have the responsibility to set that value?
If a block contains multiple applications which are managed by different Pipeds, so which piped will mark that `runnable field for the next block?

The control plane can run a simple check every time based on the deployment status to decide?
How do you think?

Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: > who will have the responsibility to set that value?

control-plane will do it, when pipedx submit its deployment state (as successes), we will check for state of all node in its block, if the block's nodes are all completed, mark the runnable state for next block as true, that's my plan. How do you think?

Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: > The control plane can run a simple check every time based on the deployment status to decide?

Yeh, the control-plane run the check, but the pipedx who deploy the deployment just internally check does it should start trigger its deployment, so we need that runnable lock, I think

Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: I got your points. Looks nice.

Just a small thing, about where to put runnable field, if that field will be set by control-plne so I still prefer to put it in each node. By that way we still be abe to get what we want and beside that we can have more control on running order of each application. I mean controlling an application gives us more opportunity than controlling whole block.
(example, we allow apps in a block could be run serially.)

Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: okay, I get that. For now, it will be lock/unlock all nodes of a block at the same time, right?

Comment (khanhtc1202) [reply pkg/app/api/service/pipedservice/service.proto:491]: note: we may don't need that lock, since the control-plane knows when the previous block deployed successfully, one rpc call to ask control-plane whether the pipedx should start its deployment is enough 👍

Comment (nghialv) [reply pkg/app/api/service/pipedservice/service.proto:491]: Cool. 👍

File: pkg/app/api/grpcapi/piped_api.go (hunk 2 above)
Comment (nghialv) [reply pkg/app/api/grpcapi/piped_api.go]: How do you think about adding a chain option to `CreateDeploymentRequest` and initializing chain commands in that API instead of creating this new `CreateDeploymentChain`?

In that way,  we can have both chain ID and the first deployment ID at the same time.

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go]: So you mean to make `CreateDeploymentRequest` has the ability to make chain commands to trigger all other applications in the chain and the RPC call to create deployment for the first application will do it? In that case, I wonder 2 things:
- we also need to provide matchers to filter out applications for chain block in that request body, right?
- though it's named `CreateDeployment`, it may also create `Command` in some conditions.

Besides, not sure what do you mean by "we can have both chain ID and the first deployment ID at the same time", since we separate the RPC to create the first deployment of the chain with the deployment chain itself, we have to first: create the deployment chain and wait until it's succeeded, then: use the deployment chain id created to find the deployment chain object in the datastore and add the on going create deployment as the deploymentChain.block[0] item, right?
Currently, that part is addressed here: https://github.com/pipe-cd/pipe/pull/2815/files#diff-dc8d28c3b055fcad0f022406b744ce756e3be8c630b6c40aef12e026c6a58dcfR1019-R1032

Comment (nghialv) [reply pkg/app/api/grpcapi/piped_api.go]: > we also need to provide matchers to filter out applications for chain block in that request body, right?

Yes, we will need those `matchers`.

> not sure what do you mean by "we can have both chain ID and the first deployment ID at the same time"

I meant by that way, we can have chain ID, deployment objects, commands in the same API. Then the first deployment can have its needed chain ID seamlessly.

---

What I am concerned about is how to set the chain ID for the first deployment object. In the case of doing with separate RPCs, the flow will be like this, right?

- Piped generates chain ID locally
- Piped requests to create a new Deployment with the generated chain ID
- Piped requests to create a new Chain (commands, etc...)

IMO, generating the chain ID locally is not good. A Piped may send a chain ID of other projects (accidentally or intentionally).

So I think it would be better to avoid that. Another idea is reordering the above steps to

- Piped requests to create a new Chain (commands, etc..) and chain ID is returned
- Piped requests to create a new Deployment with the received chain ID

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go]: Yeh, in that case, we have to use the chain id to find the chain and add deployment ref to chain: we have 2 different possible cases:
 + the common case: the creating deployment come by pipedx when it's triggered by chain_sync_app command, and besides with the chain id, we also need block index to add the deployment to the right block
 + the uncommon case: the creating deployment is the first deployment of the chain, in this case, we have to add a way to separate it with others, maybe by check for `matcher` existed in the request body or not, or specified block_index = 0 in the request body

> Piped requests to create a new Chain (commands, etc..) and chain ID is returned
One minor point for this is, at the time when we make commands chain_sync_app for applications in the chain, the first application deployment is not yet in the block[0] of the deployment chain model, if pipedx get the command and ask (via RPC) controlplane should it handle its command, we don't have any "previous" stage to determine. To avoid that, we have to ensure commands only be created after the first deployment is available as chain block[0] by making it as currently or adding `matcher` to the CreateDeploymentRequest as optional.

If you're worried about the ChainId generated in the client (piped side), we have another way to avoid that. That is generate it in the controlplane (API side) and update this `req.FirstDeployment` [here](https://github.com/pipe-cd/pipe/pull/2815/files#diff-dc8d28c3b055fcad0f022406b744ce756e3be8c630b6c40aef12e026c6a58dcfR1061) before actually create/add it to the datastore. How do you think about that?

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1049]: Addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/4a4ec2147c4103c411e5fe56a46d72334dd553fa

File: pkg/app/api/grpcapi/piped_api.go [hunk 4]
```diff
@@ -25,6 +25,8 @@ import (
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 
+	"github.com/google/uuid"
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:28]: Should be in the same import group.

File: pkg/app/api/grpcapi/piped_api.go [hunk 5]
```diff
@@ -963,6 +967,129 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
...
+				},
+			})
+		}
+
+		return nodes, apps, nil
+	}
+
+	chainBlocks := make([]*model.ChainBlock, 0, len(req.Matchers)+1)
+	// Add the first deployment which created by piped as the first block of the chain.
+	chainBlocks[0] = &model.ChainBlock{
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:1019]: Have to use `append` to avoid panic because the size of `chainBlocks` was initialized by 0.

File: pkg/app/api/grpcapi/piped_api.go [hunk 6]
```diff
@@ -963,6 +967,129 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
...
+	// Make sync application command for applications of the chain.
+	for _, app := range apps {
+		cmd := model.Command{
+			Id:            uuid.New().String(),
+			PipedId:       app.PipedId,
+			ApplicationId: app.Id,
+			ProjectId:     app.ProjectId,
+			Commander:     dc.Id,
+			// TODO: Add new command type in order to separate this in chain sync command with other commands.
+			Type: model.Command_SYNC_APPLICATION,
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:1077]: Should we use the newly created command instead of this?

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1077]: Will address with a separated PR 🙏

File: pkg/app/api/grpcapi/piped_api.go (hunk 5 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go]: Nice catch, addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/8790eda637f06196a46a23b46f044bc95d1bb458

File: pkg/app/api/grpcapi/piped_api.go (hunk 4 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go]: Addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/8790eda637f06196a46a23b46f044bc95d1bb458

File: pkg/app/api/grpcapi/piped_api.go (hunk 6 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:1076]: Changed my mind, lets me address it in this PR

File: pkg/app/api/grpcapi/piped_api.go [hunk 7]
```diff
@@ -963,6 +966,129 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
 	return nil, status.Errorf(codes.Unimplemented, "ReportUnregisteredApplicationConfigurations is not implemented yet")
 }
 
+// CreateDeploymentChain creates a new deployment chain object and all required commands to
+// trigger deployment for applications in the chain.
+func (a *PipedAPI) CreateDeploymentChain(ctx context.Context, req *pipedservice.CreateDeploymentChainRequest) (*pipedservice.CreateDeploymentChainResponse, error) {
+	projectID, _, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:975]: Add a check to ensure that the application of the requested deployment belongs to that this Piped or not.

File: pkg/app/api/grpcapi/piped_api.go (hunk 6 above)
Comment (nghialv) [reply pkg/app/api/grpcapi/piped_api.go:1076]: Sure.

File: pkg/app/api/grpcapi/piped_api.go [hunk 8]
```diff
@@ -963,6 +966,129 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
...
+			Id:            uuid.New().String(),
+			PipedId:       app.PipedId,
+			ApplicationId: app.Id,
+			ProjectId:     app.ProjectId,
+			Commander:     dc.Id,
+			// TODO: Add new command type in order to separate this in chain sync command with other commands.
+			Type: model.Command_SYNC_APPLICATION,
+			SyncApplication: &model.Command_SyncApplication{
+				ApplicationId: app.Id,
+				SyncStrategy:  model.SyncStrategy_AUTO,
```
Comment (nghialv) [inline pkg/app/api/grpcapi/piped_api.go:1079]: ChainID and BlockIndex will be added in separate PR, right?

File: pkg/app/piped/trigger/trigger.go [hunk 9]
```diff
@@ -269,15 +270,53 @@ func (t *Trigger) checkRepoCandidates(ctx context.Context, repoID string, cs []c
 			strategy = model.SyncStrategy_AUTO
 		}
 
-		// Build deployment model and send a request to API to create a new deployment.
-		deployment, err := t.triggerDeployment(ctx, app, appCfg, branch, headCommit, commander, strategy, strategySummary)
+		// TODO: Add ability to get deployment chain id from CHAIN_SYNC_APPLICATION command.
+		var deploymentChainId string
```
Comment (nghialv) [inline pkg/app/piped/trigger/trigger.go:274]:
```suggestion
		var deploymentChainID string
```

File: pkg/app/piped/trigger/trigger.go [hunk 10]
```diff
@@ -269,15 +270,53 @@ func (t *Trigger) checkRepoCandidates(ctx context.Context, repoID string, cs []c
...
 			t.notifyDeploymentTriggerFailed(app, appCfg, msg, headCommit)
 			t.logger.Error(msg, zap.Error(err))
 			continue
 		}
 
+		// In case the triggered deployment is of application that can trigger a deployment chain
+		// create a new deployment chain with it's configuration.
+		if appCfg.PostSync != nil && appCfg.PostSync.DeploymentChain != nil {
+			if err := t.triggerDeploymentChain(ctx, appCfg.PostSync.DeploymentChain, deployment); err != nil {
+				msg := fmt.Sprintf("failed to trigger new deployment chain: %v", err)
```
Comment (nghialv) [inline pkg/app/piped/trigger/trigger.go:298]:
```suggestion
				msg := fmt.Sprintf("failed to trigger application %s and its deployment chain: %v", app.Id, err)
```

File: pkg/app/api/grpcapi/piped_api.go (hunk 8 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go]: addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/91a442b023c4f1a19c95b4cdc36c5c35d8c3b0c9

File: pkg/app/api/grpcapi/piped_api.go (hunk 7 above)
Comment (khanhtc1202) [reply pkg/app/api/grpcapi/piped_api.go:976]: addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/91a442b023c4f1a19c95b4cdc36c5c35d8c3b0c9
